### PR TITLE
cookbook: multiclass logistic regression

### PR DESCRIPTION
**Pull request by OXPHOS**

### CODE REVIEW

**Comment by karlnapf** [comment]
Need integration test data (and I guess remove the old example, or tell us that there does not exist one)

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst [hunk 1]
```diff
@@ -0,0 +1,43 @@
+===============================
+Multi-class Logistic Regression
+===============================
+
+Logistic regression is a binary classifier. The samples are predicted to be one case or not based on the probability calculated by feature matrix and logistic function:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: Logistic regression is a binary classifier that assigns class probabilities as

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+.. math::
+ 
+    z = \alpha +  \mathbf{\beta x} = \alpha + \beta_{1}x_{1} + ... + \beta_{d}x_{d}
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: You have to say what these letters mean, otherwise the math is not helpful.
Also, if you could state the loss function that is minimised, that would be great

**Comment by karlnapf** [comment]
Needs some minor cleanups and integration test data and then is good to go

**Comment by OXPHOS** [comment]
@karlnapf updated

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst [hunk 3]
```diff
@@ -0,0 +1,51 @@
+===============================
+Multi-class Logistic Regression
+===============================
+
+Logistic regression is a binary classifier that assigns class probabilities as: 
+
+.. math::
+ 
+    \mathrm{z} = \alpha +  \mathbf{\beta x} = \alpha + \beta_{1}x_{1} + ... + \beta_{d}x_{d} = \mathbf{\theta}^{T}\mathbf{x}
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: See [equation 17](http://yelab.net/software/SLEP/manual.pdf) for the actual loss minimised.
You can translate that to this problem.

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    
+.. math::
+
+    p = 1 / (1 + e^{-\mathrm{z}})
+    
+where it maps a sample :math:`\mathbf{x}` in :math:`d`-dimensional feature space to :math:`p`, the probability one sample falls in one class.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: there is not feature space here

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+The loss function that needs to be minimized is:
+
+.. math::
+
+    {\min_{\mathbf{x}}}\sum_{l=1}^{k}\sum_{i=1}^{m}w_{il}log(1+exp(-y_{il}(\mathbf{x}_l^Ta_{il} + c_l))) + \lambda\left \| \mathbf{x} \right \|_{l_1/l_q}
+    
+where :math:`a_{il}` denotes the :math:`i`-th sample for the :math:`l`-th class, :math:`w_{il}` is the weight for :math:`a_{il}^T`, 
+:math:`y_{il}` is the response of :math:`a_{il}`, and :math:`c_l` is the intercept (scalar) for the :math:`l`-th class.
+
```
Comment (OXPHOS) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: @karlnapf maybe you want to check again? This one is complicated..

Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: Sure :)
No worries if this takes a bit. It *is* complicated.
The math is not quite synchronised. Maybe this one helps? [link1](http://stats.stackexchange.com/questions/31714/logistic-regression-for-multiclass) [link2](http://blog.datumbox.com/machine-learning-tutorial-the-multinomial-logistic-regression-softmax-regression/)

You basically need to state two things
1) How the probability for a data x to be in class y is computed p(y_i=k | x_i)
2) The above will involve a parameter vector \theta. You need to state the loss that is minimised over \theta (not over the data x itself btw). 
3) The loss will be regularised, which has different implications. The slep manual I posted contained the way it is done here.

Currently there are some problems.
 * min is over x, which doesnt make sense, it should be over \theta
 * the p in first line is only for one class

Just combine all pieces of information and then we will have it running :)

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+The loss function that needs to be minimized is:
+
+.. math::
+
+    {\min_{\mathbf{x}}}\sum_{l=1}^{k}\sum_{i=1}^{m}w_{il}log(1+exp(-y_{il}(\mathbf{x}_l^Ta_{il} + c_l))) + \lambda\left \| \mathbf{x} \right \|_{l_1/l_q}
+    
+where :math:`a_{il}` denotes the :math:`i`-th sample for the :math:`l`-th class, :math:`w_{il}` is the weight for :math:`a_{il}^T`, 
+:math:`y_{il}` is the response of :math:`a_{il}`, and :math:`c_l` is the intercept (scalar) for the :math:`l`-th class.
+    
+We apply logistic regression to multi-class prediction by using one-vs-.rest multi-class strategy.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: It is actually true multiclass. No one-vs-rest strategies

@lisitsyn

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst (hunk 5 above)
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: for classes maybe use  K for the number of classes and k as index

**Comment by karlnapf** [comment]
travis fails, but if you rebase that might go

**Comment by OXPHOS** [comment]
@karlnapf updated

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst [hunk 7]
```diff
@@ -0,0 +1,49 @@
+===============================
+Multi-class Logistic Regression
+===============================
+
+Multinomial logistic regression assigns the sample :math:`\mathbf{x}_i` to class :math:`c`
+based on :math:`P`, the probability for sample :math:`\mathbf{x}_i`: to be in class :math:`c`:
+
+.. math::
+
+    P(Y_i = c | \mathbf{X}_i) = \frac{\mathrm{exp}(\mathbf{\theta}^T_c\mathbf{X}_i)}{1+ \sum_{k=1}^{K}\mathrm{exp}(\mathbf{\theta}^T_k\mathbf{X}_i)}
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: btw if you locally run `make cookbook` then you can look at the rendered html output (See readme).
If you do this, you can spot latex errors in this one (it is not rendered properly), as also visible [here](http://buildbot.shogun-toolbox.org/cookbook_pr/583fd0d3f73ac5166e680e4a7f574aed41ee688b/examples/classifier/multiclass_logisticregression.html)

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst [hunk 8]
```diff
@@ -0,0 +1,49 @@
+===============================
+Multi-class Logistic Regression
+===============================
+
+Multinomial logistic regression assigns the sample :math:`\mathbf{x}_i` to class :math:`c`
+based on :math:`P`, the probability for sample :math:`\mathbf{x}_i`: to be in class :math:`c`:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: remove P here

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst (hunk 7 above)
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: use `\log` and `\exp` latey macros

Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: use `^top` for transposed vectors

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+    P(Y_i = c | \mathbf{X}_i) = \frac{\mathrm{exp}(\mathbf{\theta}^T_c\mathbf{X}_i)}{1+ \sum_{k=1}^{K}\mathrm{exp}(\mathbf{\theta}^T_k\mathbf{X}_i)}
+
+in which K is the number of classes.
+
+The loss function that needs to be minimized is:
+
+.. math::
+
+    {\min_{\mathbf{\theta}}}\sum_{k=1}^{K}\sum_{i=1}^{m}w_{ik}log(1+mathrm{exp}(-y_{ik}(\mathbf{x}_l^T\mathbf{a}_{ik} + c_k))) + \lambda\left \| \mathbf{x} \right \|_{l_1/l_q}
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: same here for exp and log

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+where :math:`a_{ik}` denotes the :math:`i`-th sample for the :math:`k`-th class, :math:`w_{ik}` is the weight for :math:`a_{ik}^T`,
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: please be consistent with boldface math. This a is not bold, whereas the above one is

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+:math:`y_{ik}` is the response of :math:`a_{ik}`, and :math:`c_k` is the intercept (scalar) for the :math:`k`-th class.
+
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: Great, this is much better now. It can stay like that once you cleared up these minor things

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst (hunk 9 above)
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: You have that l_1/l_q from the manual I guess?
Do they say more about that? If so, also mention it below

File: doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+-------
+Example
+-------
+
+Imagine we have files with training and test data. We create CDenseFeatures (here 64 bit floats aka RealFeatures) and :sgclass:`CMulticlassLabels` as
+
+.. sgexample:: multiclass_logisticregression.sg:create_features
+
+We create an instance of the :sgclass:`CMulticlassLogisticRegression` classifier by passing it the dataset, lables, and specifying the regularization constant for each machine
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_logisticregression.rst]: I guess regularization constant is lambda? Mention it

**Comment by karlnapf** [comment]
Great, this is almost there. Travis is already happy (should be great after a rebase)
Some minor edits in latex required and then can be merged

**Comment by OXPHOS** [comment]
There's not much info about l1/lq regularization. Feels like it's some basic knowledge everyone knows..

**Comment by karlnapf** [comment]
ok leave it then. 
Can you add integration test data ans squash and ping me once everything is there?

**Comment by OXPHOS** [comment]
@karlnapf data version updated and squashed. Ready to merge.

**Comment by karlnapf** [comment]
Phew, that was a long one. But great that it is merged!
